### PR TITLE
flake-report: stop re-filing issues closed with no new failures

### PR DESCRIPTION
**Pull request by aditya-shantanu**
## What this fixes

`dev/tools/flake-report --update-issues` dedups findings against **open** `kind/flake` issues only (`find_existing_issue` over the open-issue list by the `<!-- flake-report:... -->` marker). When a maintainer closes an issue as a verified false positive while its stale failing columns are still inside the 100-run TestGrid window, the next nightly run finds no open match and files a duplicate.

This has already happened in practice:

- #1631 was closed as a verified non-issue, and the next run re-filed it as #1640.
- #1662, #1663, and #1664 were closed, and re-appeared as #1692, #1693, and #1694.

The only workaround has been to leave verified non-issues open purely as re-file suppressants, which pollutes the `kind/flake` backlog.

## The fix

When no open issue matches a finding's marker, `update_issues` now also checks closed `kind/flake` issues for the same marker (most recently closed match wins, marker-only — no title fallback):

- If the finding's `last_failure_ts` is **not newer** than the closed issue's `closedAt`, filing is skipped and logged in the existing skip-log style: `skip (closed #N, no failures since close): <test>`. Those failures are the same stale columns the closer already judged.
- If there **are** failures strictly newer than the close, a fresh issue is filed as before — genuine recurrences must still surface — and the new body links the prior issue number for lineage.
- An unparsable or missing `closedAt` files rather than suppresses.
- The same logic applies to both per-test flaky issues and infra-tab issues (infra findings carry `last_failure_ts` too).

Open issues take precedence exactly as before; the closed-issue path only runs when there is no open match.

This unblocks closing verified false positives immediately instead of waiting for the stale columns to age out of the TestGrid window.

## Tests

Extended `dev/tools/flake_report_test.py` following the existing injected-stub pattern: pure-helper tests for `should_skip_for_closed` and `find_closed_issue`, plus `update_issues` decision-path tests with a stubbed `gh` covering skip/file for both per-test and infra findings and open-issue precedence. `python3 dev/tools/flake_report_test.py` green (34 tests); all other `dev/tools/*_test.py` suites still pass.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved flaky-test and infrastructure issue tracking by preventing duplicate reports when a matching issue was already closed after the latest failure.
  * New reports are linked to relevant previously closed issues when a newer failure occurs.
  * Improved matching accuracy by relying on issue body markers.

### CODE REVIEW

File: dev/tools/flake-report [hunk 1]
```diff
@@ -460,6 +460,45 @@ def find_existing_issue(existing_issues, marker, test_name=None):
     return None
 
 
+def find_closed_issue(closed_issues, marker):
+    """Finds the most recently closed issue that tracked this marker, or None.
+
+    Only the HTML marker is matched (no title fallback): a closed issue may
+    only suppress re-filing for the exact finding it tracked.
+    """
+    matches = [i for i in closed_issues if marker in i.get("body", "")]
```
Comment (janetkuo) [inline dev/tools/flake-report:469]: `i.get("body", "")` returns `None` if an issue has `"body": null` in the GitHub API JSON response (e.g. issues created without a description), crashing `update_issues`. Add a test to cover this case too.

File: dev/tools/flake-report [hunk 2]
```diff
@@ -460,6 +460,45 @@ def find_existing_issue(existing_issues, marker, test_name=None):
...
+    just resurrect the closed issue. Only a failure strictly newer than
+    `closed_at` (ISO-8601, from GitHub) is evidence of a genuine recurrence.
+    An unparsable or missing close time files rather than suppresses.
+    """
+    if not closed_at:
+        return False
+    try:
+        closed_ms = datetime.fromisoformat(
+            closed_at.replace("Z", "+00:00")).timestamp() * 1000
+    except ValueError:
```
Comment (janetkuo) [inline dev/tools/flake-report]: `should_skip_for_closed` catches only `ValueError`. If GitHub returns a non-string or malformed field for `closed_at`, `.replace()` can raise `AttributeError` or `TypeError`. Catching `(ValueError, TypeError, AttributeError)` aligns with the documented intent that unparsable close times file rather than suppress.

File: dev/tools/flake-report [hunk 3]
```diff
@@ -460,6 +460,45 @@ def find_existing_issue(existing_issues, marker, test_name=None):
...
+        return False
+    return last_failure_ts <= closed_ms
```
Comment (janetkuo) [inline dev/tools/flake-report]: When TestGrid run timestamps are missing or cannot be parsed, `last_failure_ts` defaults to `0`. Comparing `0 <= closed_ms` evaluates to `True`, causing recurring flakes with missing timestamps to be suppressed indefinitely against prior closed issues. If `last_failure_ts` is `0` or missing, we lack evidence that the failure preceded the close, so it should not be suppressed.

File: dev/tools/flake-report [hunk 4]
```diff
@@ -540,7 +589,16 @@ def update_issues(repo, flaky, infra, dry_run):
...
+            prior = find_closed_issue(closed, marker)
+            if prior and should_skip_for_closed(
+                    prior.get("closedAt"), f["last_failure_ts"]):
+                actions.append(
+                    f"skip (closed #{prior['number']}, no failures since "
+                    f"close): {f['short']}")
+                continue
             body = issue_body_for_flake(f, marker) + f"\n<!-- last-reported-failure={f['last_failure_ts']} -->\n"
+            if prior:
+                body += lineage_note(prior)
```
Comment (janetkuo) [inline dev/tools/flake-report]: Adding `lineage_note` only during issue creation causes the reference to be lost on subsequent runs. When an issue is re-filed and later updated with newer failures (lines 581–590), `body` is regenerated from `issue_body_for_flake(...)` and passed to `gh issue edit --body`, overwriting the `Previously tracked in #...` note. The lineage note should be retained when editing or preserved from the existing issue body.

File: dev/tools/flake_report_test.py [hunk 5]
```diff
@@ -352,5 +353,182 @@ def test_render_report_calls_out_benign_red_runs(self):
...
+
+    def test_matches_marker_only_not_title(self):
+        issues = [
+            {"number": 1, "title": "[FLAKE] TestA (tab)", "body": "no marker",
+             "closedAt": "2026-01-05T00:00:00Z"},
+            {"number": 2, "body": f"{self.MARKER}\nbody",
+             "closedAt": "2026-01-01T00:00:00Z"},
+        ]
+        self.assertEqual(
+            flake_report.find_closed_issue(issues, self.MARKER)["number"], 2)
```
Comment (janetkuo) [inline dev/tools/flake_report_test.py]: Please add a test case covering an issue entry where `"body": None` to ensure `find_closed_issue` safely handles issues without descriptions without raising a `TypeError`.

**Comment by aditya-shantanu** [comment]
Live evidence for reviewers: this morning's pipeline run re-filed 8 duplicates (#1701–#1708) of verified-and-closed issues — every one with zero failures postdating the close, exactly the loop this PR eliminates. With this merged, today's run would have logged 8 skips instead.

**Comment by aditya-shantanu** [comment]
Second consecutive day of re-file churn: today's run filed 8 more duplicates (#1711–#1718, all zero-new-failures, all closed with pointers) — 16 duplicate issues in two days. @janetkuo @igooch PTAL when you can; this is a small, well-tested change to dev/tools/flake-report and it ends this loop.

**Review by janetkuo — APPROVED** [review]
/lgtm

https://github.com/kubernetes-sigs/agent-sandbox/pull/1696/changes#r4041069511 can be fixed in a follow up
